### PR TITLE
[inductor] Add ir.WelfordReduction with multiple outputs

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104725

This replaces `var_unnormalized` reduction type with `welford_reduce` which takes the input data and outputs not just the variance, but also the mean and weights which account for the full welford accumulator state. Thus we can avoid re-computing the mean, and we now have enough information to create a multilayer reduction which I implement here by adding a second reduction type called `welford_combine` which reduces over all three inputs simultaneously.

Multi-layer support is particularly important as normalization operators like BatchNorm are being split in many timm models, which meant `var_unnormalized` had to fall back to two-pass variance calculation. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/lib/libshm/CMakeLists.txt [hunk 1]
```diff
@@ -29,7 +29,12 @@ set_target_properties(shm PROPERTIES
   PREFIX "lib"
   IMPORT_PREFIX "lib"
   CXX_STANDARD 17)
-target_link_libraries(shm PUBLIC torch)
+target_link_libraries(shm PUBLIC
+  torch
+  CUDA::cupti
+  CUDA::cusparse
+  ${CUDNN_LIBRARY_PATH}
+)
```
Comment (lezcano) [inline torch/lib/libshm/CMakeLists.txt:37]: ooops

Comment (lezcano) [reply torch/lib/libshm/CMakeLists.txt:37]: https://www.youtube.com/watch?v=CduA0TULnow

**Comment by peterbell10** [comment]
This gives a nice 10% speedup in many timm models:
https://hud.pytorch.org/benchmark/timm_models/inductor_no_cudagraphs?startTime=Wed%2C%2026%20Jul%202023%2014%3A01%3A20%20GMT&stopTime=Wed%2C%2002%20Aug%202023%2014%3A01%3A20%20GMT&granularity=hour&mode=training&dtype=amp&lBranch=gh/peterbell10/582/orig&lCommit=1ee353815cda0a504348075fc81c8f3586ab75ff&rBranch=main&rCommit=0af3203c728278920f6e64b22f63181c6f88628a

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -1079,6 +1086,289 @@ def intermediate_fn(index, reduction_index):
         )
 
 
+def num_reduction_outputs(reduction_type):
+    return 3 if "welford" in reduction_type else 1
+
+
+class WelfordReduction(Reduction):
```
[the remark this answers is not in the document]
Comment (lezcano) [reply torch/_inductor/ir.py:1093]: Wouldn't it be possible to generalise `Reduction` to be able to handle multiple inputs / outputs? It's a bit concerning how much repeated code we have here...

**Review by lezcano — APPROVED** [review]
I'm not the biggest fan of all the duplicate code in ir.py. It'd be nice to have a go at some point at having a Reduction IR node that handles single and multiple outputs.

At any rate, great PR twisting inductor's arm to make it support this reduction via the `store_reduction` split.

File: torch/_inductor/index_propagation.py [hunk 3]
```diff
@@ -171,15 +174,22 @@ def unwrap(self, a: Union[Any, IndexPropVar]) -> Any:
...
         # Fallback to the wrapped handler
         new_args = [self.unwrap(a) for a in args]
         new_kwargs = {k: self.unwrap(v) for k, v in kwargs.items()}
-        return IndexPropVar(getattr(self._inner, name)(*new_args, **new_kwargs))
+        return self.wrap(getattr(self._inner, name)(*new_args, **new_kwargs))
 
     def propagate_sympy(
         self, name: str, args: Tuple, kwargs: Dict[str, Any]
-    ) -> Union[TypedExpr, IndexPropVar]:
+    ) -> IndexPropVar:
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply torch/_inductor/index_propagation.py:192]: This function never returns a tuple at the moment, so the typing here is correct.

File: torch/_inductor/codegen/triton.py [hunk 4]
```diff
@@ -1407,6 +1412,16 @@ def reduction(self, dtype, src_dtype, reduction_type, value):
         reduction_sizes = ["None" for _ in self.range_trees]
         reduction_sizes[-1] = ":"
 
+        # The shape of the reduction dimension changes the output, so
+        # explicitly broadcast the input to the expected shape.
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply torch/_inductor/codegen/triton.py]: Added an example that should be clearer. Before this change
```python
tmp0 = ops.constant(1, torch.int64)
tmp1 = ops.reduction(torch.int64, torch.int64, "sum", tmp0)
```
becomes
```python
tmp0 = tl.full([1, 1], 1, tl.int64)
tmp1 = tl.sum(tmp0, 1)[:, None]
```
which is wrong because summing over 1 element is not equivalent to summing over `RBLOCK` elements.
